File: multithreaded_script.py
import numpy as np
import random
import time
from tqdm import tqdm
import concurrent.futures

start_time = time.time()
param_list = np.linspace(0,1,1000)
iterations =1000
op_arr = np.array([])

def process_param(i):
    arr = []
    for _ in range(iterations):
        arr.append(random.choices([0,1],[param_list[i],(1-param_list[i])]))
    title_str = "Data/data_param=" + str(param_list[i]) + "_iterations=" + str(iterations) +".txt"
    np.savetxt(title_str, np.array(arr))
    return arr

with concurrent.futures.ThreadPoolExecutor() as executor:
    results = list(tqdm(executor.map(process_param, range(len(param_list))), total=len(param_list), desc="Progress"))

op_arr = np.concatenate(results)
print("--- %s Total time taken ---" % (time.time() - start_time))
print('Done')

File: multithreaded_script.jl

using Distributed

addprocs(4)

@everywhere begin
    using ProgressMeter
    using Random, RandomNumbers, LinearAlgebra, Statistics, LoopVectorization
    using DelimitedFiles, Printf, StatsBase, Distributions    

start_time = time()
param_list = range(0, 1, length=1000)

function generate_data(param)
    op_arr = []
    iterations = 1000
    for _ in 1:iterations
        prob_weights = weights([param, 1-param])
        push!(op_arr, sample(0:1, prob_weights))
    end
    title_str = "Data/data_param=$(param)_iterations=$(iterations).txt"
    writedlm(title_str, op_arr)
end

end

# Use pmap to apply the function to each element of param_list in parallel
@showprogress 1 "Progress:" pmap(generate_data, param_list)

println("--- $(time() - start_time) Total time taken ---")
println("Done")